pytensor/tensor/subtensor.py

```diff
@@ -178,16 +215,51 @@ def as_index_literal(
             as_index_literal(idx.step),
         )
 
+    if not isinstance(idx, Variable):
+        raise NotScalarConstantError()
+
+    if isinstance(idx.type, NoneTypeT):
+        return None
+
+    if isinstance(idx, ScalarConstant):
+        return cast(int | None, idx.data)
+
+    if isinstance(idx, TensorConstant):
+        return cast(int, idx.data.item())
+
+    if isinstance(idx, SliceConstant):
+        return cast(slice, idx.data)
+
+    if isinstance(idx.type, SliceType):
+        assert idx.owner is not None
+        return slice(*map(as_index_literal, idx.owner.inputs))
+
+    # Other kinds of variables are not supported
     raise NotScalarConstantError()
 
 
 def get_idx_list(inputs, idx_list):
     return indices_from_subtensor(inputs[1:], idx_list)
 
 
+@overload
+def get_canonical_form_slice(
+    theslice: slice | SliceConstant,
```

Comment:
It's not just Constants right?
```suggestion
    theslice: slice | SliceVariable,
```

Reply:
Yes, but there's no `SliceVariable` type at the moment

Reply:
Oh strange. So there are Variables with type SliceType but no specific class like TensorVariable or ScalarVariable?

Reply:
exactly